modules/solid_properties/include/userobjects/ThermalSolidProperties.h

```diff
@@ -0,0 +1,115 @@
+    DualReal result = x;                                                                           \
+    result.derivatives() = T.derivatives() * dxdT;                                                 \
+    return result;                                                                                 \
+  }                                                                                                \
+                                                                                                   \
+  virtual void want##_from_T(const DualReal & T, DualReal & val, DualReal & d##want##dT) const     \
+  {                                                                                                \
+    solidPropError(__PRETTY_FUNCTION__, " derivative derivatives not implemented.");               \
+    Real dummy, tmp;                                                                               \
+    val = want##_from_T(T);                                                                        \
+    want##_from_T(T.value(), dummy, tmp);                                                          \
+    d##want##dT = tmp;                                                                             \
+  }
+
+/**
+ * Adds function definitions with not implemented error. These functions should be overriden in
+ * derived classes where required. AD versions are constructed automatically using propfuncAD.
+ */
+#define propfunc(want)                                                                             \
+  virtual Real want##_from_T(const Real &) const                                                   \
+  {                                                                                                \
+    mooseError(__PRETTY_FUNCTION__, " not implemented.");                                          \
+  }                                                                                                \
+                                                                                                   \
+  virtual void want##_from_T(const Real & T, Real & val, Real & d##want##dT) const                 \
+  {                                                                                                \
+    solidPropError(__PRETTY_FUNCTION__, " derivatives not implemented.");                          \
+    d##want##dT = 0;                                                                               \
+    val = want##_from_T(T);                                                                        \
+  }                                                                                                \
+                                                                                                   \
+  propfuncAD(want)
+
+/**
+ * Common class for solid properties that are a function of temperature
+ */
+class ThermalSolidProperties : public SolidProperties
```

Comment:
I'm fine with either approach - it depends on how complicated we think this is going to get. 

If we end up having a lot of interfaces, it could be confusing for someone adding a new material (say, thermal properties for nickel) to know which of the many functions in `SolidProperties.h` they would need to declare. Of course they'd figure it out eventually when `ThermalSolidPropertiesMaterial` tries to call a non-existent function, so it just depends on preference or the robustness of documentation/expecting the user to open up `ThermalSolidPropertiesMaterial.C` to see exactly what functions are called. In BISON, I believe that they currently split up materials based on thermal vs. solid mechanics (i.e. there's not just one material called `StainlessSteel` that computes both the thermal conductivity and a Young's modulus. So that design choice could be interpreted as a multiple-base-class route.

But I'm also not opposed to the "put everything in one base class" approach either. This approach would be preferred if there's not an easy-to-define set of interfaces (like thermal properties, solid mechanics properties, etc.).

Reply:
In general if there is a design that avoids unimplemented functions (e.g. `virtual` methods in base classes that just have `mooseError` for example), I would prefer that. However, there are cases where the error-out in base `virtual` method is a good design. We have the latter for functors for gradient computation because those are often unnecessary and it made too much of a burden on derived-class developers when those methods were pure virtual instead.

Reply:
As for the pure virtual vs. optional override, I think that there's nothing we can do to truly enforce that a function gets implemented, since derived-class developers can simply override and use a MOOSE error there, doing the same default implementation we'd have done in the base class. There are going to be methods that we may not care to implement, since there are non-AD methods and then the `x_from_T(ADReal, ADReal, ADReal)` method, which may never get used at all, so I agree that we shouldn't unnecessarily burden ourselves with those.